File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/utils/GLAccountGroup.java
package com.jasonzqshen.familyaccounting.core.utils;

import com.jasonzqshen.familyaccounting.core.exception.format.GLAccountGroupFormatException;

public enum GLAccountGroup {
	CASH("1000"), BANK_ACCOUNT("1010"), INVESTMENT("1060"), PREPAID("1430"), ASSETS(
			"1500"), SHORT_LIABILITIES("2000"), LONG_LIABILITIES("2700"), EQUITY(
			"3010"), SALARY("4000"), INVEST_REVENUE("4010"), COST_PURE("5000"), COST_ACCI(
			"5010");

	public static final GLAccountGroup[] BALANCE_GROUP = { GLAccountGroup.CASH,
			GLAccountGroup.BANK_ACCOUNT, GLAccountGroup.INVESTMENT,
			GLAccountGroup.ASSETS, GLAccountGroup.PREPAID };
	public static final GLAccountGroup[] Liquidity_GROUP = {
			GLAccountGroup.CASH, GLAccountGroup.BANK_ACCOUNT,
			GLAccountGroup.SHORT_LIABILITIES, GLAccountGroup.PREPAID };
	public static final GLAccountGroup[] LIABILITIES_GROUP = {
			GLAccountGroup.LONG_LIABILITIES, GLAccountGroup.SHORT_LIABILITIES };
	public static final GLAccountGroup[] REVENUE_GROUP = {
			GLAccountGroup.SALARY, GLAccountGroup.INVEST_REVENUE };
	public static final GLAccountGroup[] COST_GROUP = {
			GLAccountGroup.COST_ACCI, GLAccountGroup.COST_PURE };

	private final String _id;

	private GLAccountGroup(String str) {
		_id = str;
	}

	public String toString() {
		return _id;
	}

	/**
	 * Parse string to G/L account group
	 * 
	 * @param id
	 * @return
	 * @throws GLAccountGroupFormatException
	 */
	public static GLAccountGroup parse(String id)
			throws GLAccountGroupFormatException {
		if (id == null) {
			return null;
		}

		for (GLAccountGroup g : GLAccountGroup.values()) {
			if (g.toString().equals(id)) {
				return g;
			}
		}

		throw new GLAccountGroupFormatException(id);
	}

	/**
	 * get group
	 * 
	 * @param group
	 * @return
	 */
	public static String getDescp(GLAccountGroup group) {
		if (group == GLAccountGroup.CASH) {
			return "Cash on hand";
		} else if (group == GLAccountGroup.BANK_ACCOUNT) {
			return "Bank account";
		} else if (group == GLAccountGroup.INVESTMENT) {
			return "Investment";
		} else if (group == GLAccountGroup.PREPAID) {
			return "Prepaid";
		} else if (group == GLAccountGroup.ASSETS) {
			return "Fixed assets";
		} else if (group == GLAccountGroup.EQUITY) {
			return "Equity";
		} else if (group == GLAccountGroup.LONG_LIABILITIES) {
			return "Long liabilities";
		} else if (group == GLAccountGroup.SHORT_LIABILITIES) {
			return "Short liabilities";
		} else if (group == GLAccountGroup.SALARY) {
			return "Salary";
		} else if (group == GLAccountGroup.INVEST_REVENUE) {
			return "Revenue from investment";
		} else if (group == GLAccountGroup.COST_PURE) {
			return "Cost for purpose";
		} else if (group == GLAccountGroup.COST_ACCI) {
			return "Cost accidentlly";
		}

		return "";
	}
}

File: proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Utilities/GLAccountGroup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;

namespace WinStore_FamilyAccountingCore.Utilities
{
    public enum GLAccountGroupENUM
    {
        CASH = 0, BANK_ACCOUNT, INVESTMENT, PREPAID, ASSETS
            , SHORT_LIABILITIES, LONG_LIABILITIES, EQUITY
                , SALARY, INVEST_REVENUE, COST_PURE, COST_ACCI
    }


    public class GLAccountGroup
    {
        public readonly static string[] GROUP_VALUE = {
	("1000"), ("1010"), ("1060"), ("1430"), (
			"1500"), ("2000"), ("2700"), (
			"3010"), ("4000"), ("4010"), ("5000"), (
			"5010")};

        /// <summary>
        /// Group of account group
        /// </summary>
        public static readonly GLAccountGroupENUM[] BALANCE_GROUP = { GLAccountGroupENUM.CASH,
			GLAccountGroupENUM.BANK_ACCOUNT, GLAccountGroupENUM.INVESTMENT,
			GLAccountGroupENUM.ASSETS, GLAccountGroupENUM.PREPAID };
        public static readonly GLAccountGroupENUM[] Liquidity_GROUP = {
			GLAccountGroupENUM.CASH, GLAccountGroupENUM.BANK_ACCOUNT,
			GLAccountGroupENUM.SHORT_LIABILITIES, GLAccountGroupENUM.PREPAID };
        public static readonly GLAccountGroupENUM[] LIABILITIES_GROUP = {
			GLAccountGroupENUM.LONG_LIABILITIES, GLAccountGroupENUM.SHORT_LIABILITIES };
        public static readonly GLAccountGroupENUM[] REVENUE_GROUP = {
			GLAccountGroupENUM.SALARY, GLAccountGroupENUM.INVEST_REVENUE };
        public static readonly GLAccountGroupENUM[] COST_GROUP = {
			GLAccountGroupENUM.COST_ACCI, GLAccountGroupENUM.COST_PURE };

        private readonly GLAccountGroupENUM _id;
        public GLAccountGroupENUM Identity { get { return _id; } }

        /// <summary>
        /// constructor
        /// </summary>
        /// <param name="str"></param>
        private GLAccountGroup(GLAccountGroupENUM str)
        {
            _id = str;
        }

        /// <summary>
        /// to String
        /// </summary>
        /// <returns></returns>
        public override String ToString()
        {
            return GROUP_VALUE[(int)_id];
        }

        /// <summary>
        /// equals
        /// </summary>
        /// <param name="obj"></param>
        /// <returns></returns>
        public override bool Equals(object obj)
        {
            GLAccountGroup group = obj as GLAccountGroup;
            if (group == null)
            {
                return false;
            }
            return this.Identity == group.Identity;
        }

        /// <summary>
        /// get hash code
        /// </summary>
        /// <returns></returns>
        public override int GetHashCode()
        {
            return (int)this.Identity;
        }
        /// <summary>
        /// parse identity to GLAccountGroupENUM
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        /// <exception cref="GLAccountGroupFormatException">GL account group format exception</exception>
        /// <exception cref="ArgumentNullException">Id is null</exception>
        public static GLAccountGroupENUM ParseEnum(String id)
        {
            if (id == null)
            {
                throw new ArgumentNullException();
            }

            int i = 0;
            foreach (string str in GROUP_VALUE)
            {
                if (str.Equals(id))
                {
                    return (GLAccountGroupENUM)i;
                }

                i++;
            }
            throw new GLAccountGroupFormatException(id);
        }

        /// <summary>
        /// parse string to GLAccountGroup
        /// </summary>
        /// <param name="id"></param>
        /// <returns></returns>
        public static GLAccountGroup Parse(String id)
        {
            if (id == null)
            {
                return null;
            }
            GLAccountGroupENUM e = ParseEnum(id);
            return new GLAccountGroup(e);
        }
    }
}
